Adding Equality library

uselessgoddess opened the pull request.

uselessgoddess committed:
Library is written

uselessgoddess committed:
Fixed requires

uselessgoddess committed:
Added runtime any equality

uselessgoddess committed:
Fixed names

uselessgoddess committed:
Function is now constant

uselessgoddess committed:
Function is now const

uselessgoddess committed:
Added CMake support

uselessgoddess committed:
Added Google Tests

uselessgoddess committed:
Delete Platform.Equality.Any.h

uselessgoddess committed:
Delete Platform.Equality.Tests.cpp

uselessgoddess committed:
Delete Platform.Equality.h

uselessgoddess committed:
Delete CMakeLists.txt

uselessgoddess committed:
Added Google Tests

uselessgoddess committed:
Added Google Tests

uselessgoddess committed:
Fixed Google Tests

poul250 commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,82 @@
+            int a = 0x228;
+            int b = 0x1337;
+
+            ASSERT_TRUE(std::any(a) != std::any(b));
```
better to use `ASSERT_NE(lhs, rhs);` - it writes more debug information in case of an error

poul250 commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,82 @@
+            auto a = "i love green stuff";
+            auto b = "i love green stuff";
+
+            ASSERT_TRUE(std::any(a) == std::any(b));
```
and `ASSERT_EQ` here

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,82 @@
+
+using namespace Platform::Equality;
+
+namespace PlatformEqualityTests
```
The namespace should be `Platform::Equality::Tests`

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,82 @@
+            }
+        }
+    }
+}
```
Each file should end with an empty line.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+namespace Platform::Equality
+{
+    template<class T>
+    inline auto __to_any_equal_visitor(auto func)
```
Please rename it to `ToAnyEqualityComparer`. I think there is no need to add `__` at the start, if the user want to use it should be able to use it.

poul250 replied:
It looks like it shouldn't be used by the user, the user should use RegisterEqualityComparer. Underscore notation is reserved by the standard https://en.cppreference.com/w/cpp/keyword 
>Also, all identifiers that contain a double underscore __ in any position and each identifier that begins with an underscore followed by an uppercase letter is always reserved and all identifiers that begin with an underscore are reserved for use as names in the global namespace. See identifiers for more details.

Therefore, to denote the "privacy" of the function, I propose to do the following way - wrap it in details
```cpp
namespace Platform::Equality
{
    namespace details 
    {

        template<class T>
        inline auto ToAnyEqualityComparer(auto func) { /*...*/ }

    }  // namespace details
}  // namespace Platform::Equality
```

konard replied:
OK, we can either use `__ToAnyEqualityComparer` or use namespace `Internal` instead of `details`.
I like the namespace thing better.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+                [func](std::any a, std::any b) -> std::size_t
+                {
+                    if constexpr (std::is_void_v<T>)
+                        return func();
```
Please as `{` and `}` brackets here (do no use single line if bodies). If later it will be required to extend such code or to add something for debug it will be much easier to maintain such code.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+    #define BASE_VISITOR_REGISTER(Type) __to_any_equal_visitor<Type>([](Type a, Type b) {return a == b;})
+    #define VISITOR_REGISTER(Type, Func) __to_any_equal_visitor<Type>([](Type a, Type b) {return Func;})
+    static std::unordered_map<std::type_index, std::function<bool(std::any, std::any)>>
+    any_equal_visitor {
```
Please rename `any_equal_visitor` to `AnyEqualityComparers`.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+            BASE_VISITOR_REGISTER(double),
+            BASE_VISITOR_REGISTER(long double),
+
+            VISITOR_REGISTER(const char*, std::string(a) == std::string(b)),
```
STL has `std::equal_to` specialization for `const char *` type.
So you can use `BASE_VISITOR_REGISTER(const char*)` here.
Example:
https://replit.com/@Konard/WhitesmokeWanClient#main.cpp

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+    #undef VISITOR_REGISTER
+
+    template<class T>
+    inline void register_any_equal_visitor(auto func)
```
Please rename `register_any_equal_visitor` to `RegisterEqualityComparer`.

konard reviewed (changes requested):
Changes are required to merge the pull request.

poul250 commented on cpp/Platform.Equality.Tests/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+
+include_directories(../Platform.Equality)
```
here it is better to do the following:
in the root / cpp put another CMakeListst.txt with the following content:
```
add_subdirectory (Platform.Equality)
add_subdirectory (Platform.Equality.Tests)
```
And here do
```
target_link_libraries(Platform.Equality.Tests Platform.Equality gtest gtest_main)
```

konard replied:
We can try to put in `cpp` directory, and we will see how it goes.

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+        return std::pair<std::type_index, std::function<bool(std::any, std::any)>>
+        {
+                std::type_index(typeid(T)),
+                [func](std::any a, std::any b) -> std::size_t
```
```cpp
[func = std::move(func)](const std::any& a, const std::any& b)
```

uselessgoddess replied:
Wouldn't [func] cause a lambda-capture?

poul250 replied:
`[func]` will make a copy of the functor, ```[func = std::move(funс)]``` will move the functor

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+        };
+    }
+
+    #define BASE_VISITOR_REGISTER(Type) __to_any_equal_visitor<Type>([](Type a, Type b) {return a == b;})
```
looks like here can be used standard functors
```cpp
#define BASE_VISITOR_REGISTER(Type) __to_any_equal_visitor<Type>(std::equal_to<Type>{})

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    #define BASE_VISITOR_REGISTER(Type) __to_any_equal_visitor<Type>([](Type a, Type b) {return a == b;})
+    #define VISITOR_REGISTER(Type, Func) __to_any_equal_visitor<Type>([](Type a, Type b) {return Func;})
```
should it be?
```cpp
#define VISITOR_REGISTER(Type, Func) __to_any_equal_visitor<Type>(Func)
```

konard replied:
If [this change](https://github.com/linksplatform/Equality/pull/2#discussion_r614754863) is applied, this macro can be deleted as there will be no usages.

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+    #undef VISITOR_REGISTER
+
+    template<class T>
+    inline void register_any_equal_visitor(auto func)
```
```cpp
template<class T, class Foo>
inline void register_any_equal_visitor(Foo&& func) 
{
    any_equal_visitor.insert(__to_any_equal_visitor<T>(std::forward<Foo>(func)));
}
```

uselessgoddess replied:
I can use the move idiom, but why add another template?

poul250 replied:
One template for a type, another for functor. 
Instead of a template for a functor, you can use std::function
```cpp
template<class T>
inline void register_any_equal_visitor(std::function<bool(const T&, const&)> func) 
{
    any_equal_visitor.insert(__to_any_equal_visitor<T>(std::move(func)));
}
```

uselessgoddess replied:
Isn't that enough?
```cpp
requires (...) {func(a, b) -> std::same_as<bool>}
```

poul250 replied:
🤔

poul250 replied:
Yes, looks like this is enough

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,101 @@
+
+    #define BASE_VISITOR_REGISTER(Type) __to_any_equal_visitor<Type>([](Type a, Type b) {return a == b;})
+    #define VISITOR_REGISTER(Type, Func) __to_any_equal_visitor<Type>([](Type a, Type b) {return Func;})
+    static std::unordered_map<std::type_index, std::function<bool(std::any, std::any)>>
```
```cpp
std::function<bool(const std::any&, const std::any&)>
```

uselessgoddess committed:
Review started

uselessgoddess committed:
Added empty lines

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+                BASE_VISITOR_REGISTER(long double),
+
+                BASE_VISITOR_REGISTER(const char*),
+
```
No need for empty lines here.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+            return false;
+
+        if(!Internal::AnyEqualityComparers.contains(object.type()))
+            throw std::runtime_error(std::string("Equal function for type ").append(object.type().name()).append(" is unregistered"));
```
Add brackets to all single line if bodies.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+        if(!Internal::AnyEqualityComparers.contains(object.type()))
+            throw std::runtime_error(std::string("Equal function for type ").append(object.type().name()).append(" is unregistered"));
+
+        auto equaler = Internal::AnyEqualityComparers[object.type()];
```
`equaler` should be renamed to `comparer`

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+            return std::pair<std::type_index, std::function<bool(const std::any&, const std::any&)>>
+            {
+                    std::type_index(typeid(T)),
+                    [func = std::forward<Foo>(func)](const std::any& a, const std::any& b) -> std::size_t
```
`a` ⇾ `left`
`b` ⇾ `right`

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+
+namespace Platform::Equality
+{
+
```
No need for empty line here.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#ifndef PLATFORM_EQUALITY_ANY_H
+#define PLATFORM_EQUALITY_ANY_H
```
No need to add `_H` here and in all other `.h` file, the file later can be renamed to `.hpp`, there is no need to restrict ourselves to `.h` forever.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,108 @@
+
+    namespace Internal
+    {
+        template<class T, class Foo>
```
`T` ⇾ `TValue`
`Foo` ⇾ `TEqualityComparer`

konard reviewed (changes requested):
It can be even better.

uselessgoddess committed:
Small review

uselessgoddess committed:
Small review

uselessgoddess committed:
Very small review

uselessgoddess committed:
Fixed CMakeLists

uselessgoddess committed:
Remove some lines

uselessgoddess committed:
Global fixes

normal:
- Fixed 3 CMakeLists files
- Remove some lines from tests

strange:
- Added new template to equal_to

poul250 commented on cpp/Platform.Equality.Tests/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+
+include(FetchContent)
+FetchContent_Declare(
+        googletest
```
It's not cool to download lib in cmake ;(. GTest can be easily [found](https://cmake.org/cmake/help/v3.20/module/FindGTest.html?highlight=gtest) instead:
```cmake
find_package(GTest REQUIRED)
# ...
target_link_libraries(Platform.Equality.Tests GTest::gtest GTest::gtest_main)
```

konard replied:
We also can use `vcpkg` to make `find_package` work: https://vcpkg.readthedocs.io/en/latest/examples/installing-and-using-packages/#cmake-toolchain-file

poul250 commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        {
+            auto a = (std::string)"i love green stuff";
```
nit:
It is not cpp way, in cpp people prefer:
```cpp
std::string{"i love green stuff"};
```
or
```cpp
using namespace std::literals::string_literals;
auto a = "i love green stuff"s;
```

poul250 commented on cpp/Platform.Equality/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+target_include_directories(Platform.Equality.Library INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_compile_features(Platform.Equality.Library INTERFACE cxx_std_20)
```
A more standard way is to do it through the property:
```cmake
set_target_properties(Platform.Equality.Library PROPERTIES CXX_STANDARD 20)
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,111 @@
+                        }
+                        else
+                        {
+                            return func(std::any_cast<TValue>(left), std::any_cast<TValue>(right));
```
```cpp
return func(std::any_cast<const TValue&>(left), std::any_cast<const TValue&>(right));
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,111 @@
+    template<class T>
+    inline void RegisterEqualityComparer(auto&& func)
+    {
+        Internal::AnyEqualityComparers.insert(Internal::ToAnyEqualVisitor<T>(func));
```
```cpp
std::forward<decltype(func)>(func)
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,111 @@
+    template<typename TOther>
+#ifdef __cpp_lib_concepts
+    requires
+        requires(TOther left, TOther right)
```
```cpp
requires std::equality_comparable<TOther, TOther>
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,111 @@
+                                                .append(" is unregistered"));
+        }
+
+        auto comparer = Internal::AnyEqualityComparers[object.type()];
```
```cpp
const auto& comparer = /*...*/;
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Other.h:
```diff
@@ -0,0 +1,58 @@
+namespace Platform::System
+{
+    template <typename _Type>
+    concept IEnumerable = requires(_Type object)
```
looks like std::range

poul250 commented on cpp/Platform.Equality/Platform.Equality.Other.h:
```diff
@@ -0,0 +1,58 @@
+}
+
+template <typename _Type>
+concept std_equatable = requires(_Type left, _Type right)
```
looke like std::equality_comparable

poul250 commented on cpp/Platform.Equality/Platform.Equality.Other.h:
```diff
@@ -0,0 +1,58 @@
+
+namespace std
+{
+    template <not_std_equatable T> requires Platform::System::IEnumerable<T>
```
```cpp
template <std::range T> requires !std::equality_comparable<T>
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Other.h:
```diff
@@ -0,0 +1,58 @@
+    template <not_std_equatable T> requires Platform::System::IEnumerable<T>
+    struct equal_to<T>
+    {
+        bool operator()(const T& left, const T& right)
```
```cpp
constexpr bool operator()(const T& left, const T& right) const
```

poul250 commented on cpp/Platform.Equality/Platform.Equality.Other.h:
```diff
@@ -0,0 +1,58 @@
+    {
+        bool operator()(const T& left, const T& right)
+        {
+            auto left_begin = left.begin();
```
c++20 magic
```cpp
return std::rangs::equal(left, right);
```

uselessgoddess committed:
Fixed equality

uselessgoddess committed:
Fixed equality

uselessgoddess committed:
Small fixes

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,100 @@
+            };
+        }
+
+        #define BASE_VISITOR_REGISTER(Type) ToAnyEqualVisitor<Type>(std::equal_to<Type>{})
```
Please rename `BASE_VISITOR_REGISTER` to `EQUALITY_COMPARER`

konard commented on cpp/Platform.Equality/Platform.Equality.h:
```diff
@@ -0,0 +1,9 @@
+#define PLATFORM_EQUALITY
+
+#include "Platform.Equality.Any.h"
+#include "Platform.Equality.Other.h"
```
Please rename this file to `Platform.Equality.Range.h`

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,100 @@
+        template<class TValue, class TEqualityComparer>
+        inline auto ToAnyEqualVisitor(TEqualityComparer&& func)
+        {
+            return std::pair<std::type_index, std::function<bool(const std::any&, const std::any&)>>
```
`std::function<bool(const std::any&, const std::any&)>` is used twice in this code file, you can define using for it. I suggest `EqualityComparerFunctionType` as a name.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,100 @@
+
+        if(!Internal::AnyEqualityComparers.contains(object.type()))
+        {
+            throw std::runtime_error(std::string("Equal function for type ")
```
To make this line shorter you can use `std::format`. The format string itself can be defined as a constant.

konard reviewed (changes requested):
It can be even better.

uselessgoddess committed:
Review continued

uselessgoddess committed:
Review continued

uselessgoddess committed:
Review continued

uselessgoddess committed:
Review continued

uselessgoddess committed:
Small renaming

uselessgoddess committed:
Small fixes

uselessgoddess committed:
Small fixes

uselessgoddess committed:
Strange fixes

konard commented on cpp/Platform.Equality/Platform.Equality.h:
```diff
@@ -0,0 +1,10 @@
+#define PLATFORM_EQUALITY
+
+#include "Platform.Equality.Any.h"
+#include "Platform.Equality.Base.h"
```
Rename this file to match its contents: `Platform.Equality.EqualityComparerFunctionType.h`

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,89 @@
+        {
+            std::vector<int> a {1, 2, 3, 4};
+            std::vector<int> b {1, 2, 3, 4};
+            //EXPECT_ANY_THROW(std::any(a) == std::any(b));
```
Why did you comment this line out? The test should be working or should be removed.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,93 @@
+        if (!Internal::AnyEqualityComparers.contains(object.type()))
+        {
+            // TODO later use std::format
+            throw std::runtime_error(std::string("Equal function for type ")
```
Use `std::string{"Equal function for type "}` here.

uselessgoddess committed:
Key commit

Main changes:
- fixed equality operator (moved to std::)
- fixed explicit std::equal_to from std::any
- new features in the Tests

Secondary changes:
- added Google Benchmark

uselessgoddess committed:
Formatting fixed

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,85 @@
+        {
+            return std::pair<std::type_index, EqualityComparerFunctionType<std::any>>{
+                std::type_index(typeid(TValue)),
+                [func = std::forward<TEqualityComparer>(func)](const std::any& left, const std::any& right) -> std::size_t
```
Sholud it be
```cpp
-> bool
```
?

konard replied:
I think so.

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,85 @@
+                std::type_index(typeid(TValue)),
+                [func = std::forward<TEqualityComparer>(func)](const std::any& left, const std::any& right) -> std::size_t
+                {
+                    if constexpr (std::is_void_v<TValue>)
```
what is this check for?

konard replied:
It allows to execute func without arguments?

poul250 replied:
something like?
```cpp
RegisterEqualityComparer<void>([]() { /*...*/ });
```
Wouldn't it be better to just disallow the use of `void` on concepts? In cpp we cannot create an object of `void` type

konard replied:
How do we disallow it?

poul250 replied:
```cpp
template<class TValue, class TEqualityComparer>
  requires std::is_object_v<TValue>
inline auto ToAnyEqualVisitor(TEqualityComparer&& func)
```
^ Template will not be instantiated for non-objects

poul250 commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,85 @@
+namespace std
+{
+    template<typename TAny>
+    requires std::same_as<TAny, std::any> // FIXME mark in the documentation why you need it
```
damn templates magic, looks cool 👍

poul250 commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    namespace Range
+    {
+        namespace Internal
```
The `Internal` namespace is needed only in header files, anonymous namespaces can be used in cpp files instead - they give all symbols inside them internal linking, i.e. they will not get outside.
```cpp
namespace {
  inline constexpr int test_size = 10000;
}
void foo() {
  print(test_size);
}
```

Konard replied:
This is header-only template library. So we cannot use anonymous namespaces?

poul250 commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,126 @@
+    {
+        namespace Internal
+        {
+            inline constexpr int test_size = 10000;
```
constexpr implicitly implies `const` and `inline`, so only constexpr can be left
```cpp
constexpr int test_size = 10000;
```

uselessgoddess committed:
Small fixes

uselessgoddess committed:
Small fixes

poul250 reviewed (approved):
lgtm

konard commented on cpp/Platform.Equality/Platform.Equality.EqualityComparerFunctionType.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef PLATFORM_EQUALITY_BASE
+#define PLATFORM_EQUALITY_BASE
```
The constant name should correspond to the file name.

uselessgoddess committed:
Small fixes

uselessgoddess committed:
Small fixes

konard commented on cpp/Platform.Equality/Platform.Equality.Range.h:
```diff
@@ -0,0 +1,23 @@
+    requires(!std::equality_comparable<T>)
+    struct equal_to<T>
+    {
+        constexpr bool operator()(const T& left, const T& right) const
```
Replace with something like this: 
```cpp
    template<std::ranges::range TLeft, std::ranges::range TRight>
    requires(!std::equality_comparable_with<TLeft, TRight>)
    bool operator==(const TLeft& left, const TRight& right)
    {
        return ranges::equal(left, right);
    }
```

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    template<>
+    struct equal_to<std::any>
```
This specialization is not required.

uselessgoddess committed:
Last fixes

uselessgoddess committed:
Last fixes

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,127 @@
+#include <span>
+#include <any>
+
+namespace Platform::Equality::Benchmark
```
Please rename to `Platform::Equality::Benchmarks` (`s` added at the end). And move to separate folder `Platform.Equality.Benchmarks`. Benchmarks are not Tests.

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,127 @@
+
+namespace Platform::Equality::Benchmark
+{
+    namespace Any
```
Move to separate `Any.cpp` file

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,127 @@
+        BENCHMARK(NormalCompare);
+    }
+
+    namespace Range
```
Move to separate `Range.cpp` file.

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,130 @@
+#include <span>
+#include <list>
+
+namespace Platform::Equality
```
The namespace here should be `Platform::Equality::Tests` the same as the folder.

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+int func(auto&& a)
```
Do we really need this `func` here?

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Benchmark.cpp:
```diff
@@ -0,0 +1,127 @@
+    }
+}
+
+BENCHMARK_MAIN();
```
This line can be moved to `Main.cpp`

uselessgoddess committed:
Final fixes:
- Separating tests and benchmarks
  - Moved CmakeLists
  - Fixed import googletest and benchmark
- Range equality does not change

konard commented on cpp/Platform.Equality/Platform.Equality.Range.h:
```diff
@@ -0,0 +1,23 @@
+    {
+        constexpr bool operator()(const T& left, const T& right) const
+        {
+            return ranges::equal(left, right);
```
Add `std::` for `ranges::equal` to use the same style everywhere.

konard commented on cpp/Platform.Equality.Tests/Platform.Equality.Tests.cpp:
```diff
@@ -0,0 +1,124 @@
+            std::any b = a;
+            ASSERT_EQ(a, b);
+        }
+
```
No need for empty line here.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+#endif
```
Add an empty line at the end of the file.

konard commented on cpp/Platform.Equality.Benchmarks/Range.cpp:
```diff
@@ -0,0 +1,87 @@
+
+        BENCHMARK(STL_VectorCompare);
+    }
+}
```
Add an empty line at the end of the file.

konard commented on cpp/Platform.Equality.Benchmarks/Main.cpp:
```diff
@@ -0,0 +1,7 @@
+#include "Any.cpp"
+#include "Range.cpp"
+
+BENCHMARK_MAIN();
```
Add an empty line at the end of the file.

konard commented on cpp/Platform.Equality.Benchmarks/Any.cpp:
```diff
@@ -0,0 +1,46 @@
+
+        BENCHMARK(NormalCompare);
+    }
+}
```
Add an empty line at the end of the file.

konard commented on cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+
+
+IF_SUPPORT(Tests)
+IF_SUPPORT(Benchmarks)
```
Add an empty line at the end of the file.

konard commented on cpp/.clang-format:
```diff
@@ -0,0 +1,67 @@
+IndentPPDirectives: None
+IndentWidth: 4
+KeepEmptyLinesAtTheStartOfBlocks: true
+MaxEmptyLinesToKeep: 2
```
Change this setting to `1`.

konard commented on cpp/.clang-format:
```diff
@@ -0,0 +1,67 @@
+IndentCaseLabels: true
+IndentPPDirectives: None
+IndentWidth: 4
+KeepEmptyLinesAtTheStartOfBlocks: true
```
Change this setting to `false`.

konard commented on cpp/Platform.Equality/Platform.Equality.EqualityComparerFunctionType.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef PLATFORM_EQUALITY_EQUALITYCOMPARERFUNCTIONTYPE
+#define PLATFORM_EQUALITY_EQUALITYCOMPARERFUNCTIONTYPE
```
`PLATFORM_EQUALITY_EQUALITY_COMPARER_FUNCTION_TYPE` is better.

konard commented on cpp/Platform.Equality.Benchmarks/Range.cpp:
```diff
@@ -0,0 +1,87 @@
+            auto list2 = list1;
+
+            // std::span not have operator==
+            auto a = std::span<int>(list1); // span have Range initializer ==> std::span<int>(list.begin(), list.size())
```
Please remove these comments. Comments should not express something obvious that can be read in the documentation.

konard commented on cpp/Platform.Equality/Platform.Equality.Any.h:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    template<typename TAny, typename TOther>
+    requires std::same_as<TAny, std::any> // FIXME mark in the documentation why you need it
```
Here and in all the same places you can write a reason we are using this constraint in the comment.

konard reviewed (changes requested):
Final touches.

uselessgoddess committed:
Formatting fixes
